src/actions/build/mod.rs

```diff
@@ -0,0 +1,140 @@
+mod compiler_message_parser;
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+
+use analysis::AnalysisHost;
+use hyper::Url;
+use ls_types::Diagnostic;
+use serde_json;
+use span::compiler::DiagnosticSpan;
+
+use build::{BuildPriority, BuildQueue, BuildResult};
+use lsp_data::{NotificationMessage, PublishDiagnosticsParams};
+use server::Output;
+
+use self::compiler_message_parser::{CompilerMessageParser, FileDiagnostic, ParseError};
+
+type Results = HashMap<PathBuf, Vec<Diagnostic>>;
+
+pub struct Build {
+    analysis: Arc<AnalysisHost>,
+    build_queue: Arc<BuildQueue>,
+    previous_results: Mutex<Results>,
+}
+
+impl Build {
+    pub fn new(analysis: Arc<AnalysisHost>, build_queue: Arc<BuildQueue>) -> Self {
+        Build {
+            analysis,
+            build_queue,
+            previous_results: Mutex::new(Results::new()),
+        }
+    }
+
+    pub fn clear_previous_results(&self) {
+        let mut results = self.previous_results.lock().unwrap();
+        results.clear();
+    }
+
+    pub fn execute(&self, project_path: &Path, priority: BuildPriority, out: &Output) {

```

Comment:
Similar to the last comment, this is method is just handling the build request and so I think it should be in the action handler. It is not executing the build, it is enqueing it, and then handling the result.